### PR TITLE
Update upsample ops to support round_with_scale_factor flag

### PR DESCRIPTION
**Pull request by wonjoolee95**
Fixes https://github.com/pytorch/xla/issues/4959

Update upsample ops to support round_with_scale_factor flag 

As per https://github.com/pytorch/xla/issues/4959, new flag `support round_with_scale_factor` has been introduced to umsample_* ops. Since XLA lowers `upsample_bilinear2d` and `upsample_nearest2d`, this PR updates those ops to support this new flag.

--- 

Testing:

Locally, I can now see that the previously failing test is passing: `python test/test_nn.py TestNNDeviceTypeXLA.test_upsampling_output_size_round_with_scale_factor_True_mode_bilinear_xla`

I'll let the CI verify the rest.

### CODE REVIEW

**Review by JackCaoG — APPROVED** [review]
Thanks!

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -177,13 +177,15 @@ std::pair<XLATensorPtr, XLATensorPtr> GetBinaryOperands(
 std::vector<int64_t> GetOutputSizeWithScale(
     absl::Span<const int64_t> input_size, const c10::optional<double> scales_h,
     const c10::optional<double> scales_w,
-    const std::vector<int64_t>& output_size) {
+    const std::vector<int64_t>& output_size, bool round_with_scale_factor) {
```
Comment (vfdev-5) [inline torch_xla/csrc/aten_xla_type.cpp:180]: nit: can we make `round_with_scale_factor` default to false ?

File: torch_xla/csrc/aten_xla_type.cpp [hunk 2]
```diff
@@ -3119,9 +3121,16 @@ at::Tensor XLANativeFunctions::upsample_bilinear2d(
...
+    scaled_output_size = GetOutputSizeWithScale(
+        input_dims, scales_h, scales_w, scaled_output_size,
+        /*round_with_scale_factor=*/false);
+  }
+  if (!output_size.empty()) {
+    if (!(scaled_output_size.at(0) == output_size.at(0) &&
+          scaled_output_size.at(1) == output_size.at(1))) {
+      scaled_output_size = GetOutputSizeWithScale(
+          input_dims, scales_h, scales_w, scaled_output_size,
+          /*round_with_scale_factor=*/true);
```
Comment (vfdev-5) [inline torch_xla/csrc/aten_xla_type.cpp:3133]: Seems like that `}` on the line 3127 changes a bit the logic when the check `scaled_output_size` vs `output_size` happens.

File: torch_xla/csrc/aten_xla_type.cpp [hunk 3]
```diff
@@ -3151,8 +3160,9 @@ at::Tensor XLANativeFunctions::upsample_bilinear2d_backward(
   std::vector<int64_t> scaled_output_size =
       torch::lazy::ToVector<int64_t>(output_size);
   if ((scales_h && *scales_h != 1.0) || (scales_w && *scales_w != 1.0)) {
-    scaled_output_size = GetOutputSizeWithScale(input_size, scales_h, scales_w,
-                                                scaled_output_size);
+    scaled_output_size = GetOutputSizeWithScale(
+        input_size, scales_h, scales_w, scaled_output_size,
+        /*round_with_scale_factor=*/false);
```
Comment (vfdev-5) [inline torch_xla/csrc/aten_xla_type.cpp:3165]: I think there will be a similar problem with backward pass as in the forward.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @wonjoolee95 !
I left few comments, please let me know what do you think.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 3 above)
Comment (wonjoolee95) [reply torch_xla/csrc/aten_xla_type.cpp]: Thanks, updated!

File: torch_xla/csrc/aten_xla_type.cpp (hunk 2 above)
Comment (wonjoolee95) [reply torch_xla/csrc/aten_xla_type.cpp]: Thanks for the catch, I've moved this code into the `if` statement to keep the existing behavior. Let me know if that makes more sense.

**Comment by wonjoolee95** [comment]
> Thanks for the PR @wonjoolee95 ! I left few comments, please let me know what do you think.

@vfdev-5, thanks for the comments! I've updated the PR to address the comments, let me know if the revisions make sense.

**Review by vfdev-5 — APPROVED** [review]
@wonjoolee95 thanks! Looks good to me. 

By the way, I wonder if there are some gradcheck tests in xla to verify our updates ?
I would like to make sure that `upsample_bilinear2d_backward` is 100% correct.
For example, these test cases https://github.com/pytorch/pytorch/pull/97869/files#diff-7927db349f568afca2de9b94d74ea5c3b8cb468cb6a433d0cc1e61e65c515a36R6447-R6455 could good to run.

**Comment by wonjoolee95** [comment]
The upstream PyTorch PRs got closed, so closing ours as well.

**Pull request closed without merging**